### PR TITLE
fix(swap): block submission on an expired quote and stop the countdown lying

### LINKED ISSUE #4786: [Fix] Stale swap quote is served from cache offline and its expiry is never checked [Extension / Swap]
# [Fix] Stale swap quote is served from cache offline and its expiry is never checked [Extension / Swap]

## Problem

With a quote loaded, go offline, navigate to another screen inside the extension, and come back to Swap with the same pair and the same amount. The old quote is presented as a firm, current one — with a full countdown and an enabled Continue button. Reproduced visually.

Five things line up to produce it:

1. `gcTime` is **1 day** (`lib/ui/query/queryClientDefaultOptions.ts`), so the quote is still in memory
2. The query key is the pair plus the amount, so returning with both unchanged hits the same entry
3. The quote query sets `retry: false`, so the refetch fails at once when offline
4. react-query retains the previous `data` on a failed refetch, and `isPlaceholderData` is false — so `ToAmount` treats it as a **firm** quote rather than falling back to the indicative
5. `useRefreshSwapQuoteInterval` resets the countdown to a full 60s whenever `swapQuoteData` is truthy, so a stale quote looks freshly fetched

And nothing catches it downstream: `useSwapValidationQuery` checks amount, balance, same-asset and recipient, but **never compares the quote's `expiry` against the current time**. Continue stays enabled.

So a user can submit against a quote that is minutes or hours old, with no warning anywhere in the flow.

Note for anyone triaging this alongside #4780: it is a different mechanism. Here the amount is *unchanged*, which is exactly why the cache key matches. In #4780 the amount changes, so the key changes and the cache is never involved. Fixing this does not address that report.

Swap quotes are not persisted to `chrome.storage` — they carry no `meta.shouldPersist` — so this survives in-app navigation but not a full popup close.

## Expected Behavior

An expired quote is never presented as current. The user is told to refresh, or a refresh is forced, and submission is blocked until a live quote resolves.

## Steps to Reproduce

1. Open Swap, pick a pair, enter an amount, wait for the firm quote
2. Disconnect the network
3. Without closing the popup, navigate to another screen and back to Swap
4. Select the same pair and the same amount
5. The old payout appears immediately with no loading state, the countdown restarts at 60s, and Continue is enabled

## Solution

Validate the quote's `expiry` in `useSwapValidationQuery` and return a validation error once it has passed, which blocks Continue through the path that already exists. Separately, stop `useRefreshSwapQuoteInterval` from resetting the countdown on cached data — the timer should reflect when the quote was actually fetched, not when a component last saw data, so `dataUpdatedAt` is the right input.

Treat the expiry check as the load-bearing fix; the countdown is what makes the staleness invisible, but the enabled Continue button is what makes it dangerous.

### Must Do
- [ ] `useSwapValidationQuery` rejects a quote whose expiry has passed, with a user-facing message
- [ ] The countdown derives from when the quote was fetched, not from data merely being present
- [ ] Unit tests for an expired quote and for the countdown on rehydrated cached data

### Must NOT Do
- Don't shorten the global `gcTime` — it is shared by every query in the app
- Don't disable `placeholderData: keepPreviousData` on the quote query; it is what prevents flicker on refresh (see #4037)

## Acceptance Criteria
- [ ] `yarn check:all` succeeds
- [ ] Offline, returning to an unchanged pair and amount does not present the cached quote as current
- [ ] Continue is blocked while the displayed quote is expired
- [ ] The countdown does not restart at 60s for a quote that was not just fetched

### PR DESCRIPTION
**Pull request by Ehsan-saradar**
Closes #4786.

## What & why

With a quote loaded, going offline, navigating away inside the extension and returning to the same pair and amount re-presented the cached quote as a firm, current one — countdown full, Continue enabled.

Five things lined up:

1. `gcTime` is 1 day, so the quote is still in memory
2. The query key is the pair plus the amount, so an unchanged pair and amount hits the same entry
3. The quote query sets `retry: false`, so the refetch fails at once when offline
4. react-query retains the previous `data` on a failed refetch and `isPlaceholderData` is false, so `ToAmount` treats it as **firm** rather than falling back to the indicative
5. `useRefreshSwapQuoteInterval` reset the countdown to a full 60s whenever data was merely present

And nothing caught it downstream: `useSwapValidationQuery` checked amount, balance, same-asset and recipient, but never compared the quote against the clock.

## Expiry validation

`BoundSwapQuote` carries `expiresAt` — an absolute expiry in milliseconds, present for both native and general quotes — so the check is uniform and needs no per-provider branching. `getSwapQuoteExpiryValidationError` is a pure function taking `now`, which keeps it testable without fake timers.

Submission is blocked through the validation path that already existed, so no new wiring reaches the Continue button.

## Countdown

Clamped to the quote's own remaining validity: `min(refreshInterval, secondsUntilExpiry)`.

A live quote whose expiry outlasts the interval still counts the interval down exactly as before — this is deliberately not a rewrite of the refresh cadence. An expired one reads `0:00` instead of looking freshly fetched.

## Test plan

`yarn check:all` — lint, typecheck, i18n integrity, secrets, go, docs, duplication, complexity all green; 1962 tests across 247 files pass. The one failure is `quality:architecture`, reporting circular imports inside the gitignored `clients/extension/dist-station/` build output — identical on an untouched tree, not a finding from this branch.

8 new unit tests: expiry accepted / rejected / exactly-now, and the countdown across live, expiring, expired and missing-expiry cases.

## Notes for review

- The boundary case is deliberate: a quote expiring exactly now is rejected, since it cannot be signed and broadcast in zero time.
- `gcTime` is untouched — it is shared by every query in the app.
- `placeholderData: keepPreviousData` is untouched — it is what prevents flicker on refresh (#4037).
- Not to be confused with #4780: there the amount changes, so the cache key changes and the cache is never involved. This only reproduces with the amount unchanged.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added localized messaging for expired swap quotes in supported languages.
  * Swap forms now indicate when a quote has expired and prompt users to refresh for the latest price.

* **Bug Fixes**
  * Refresh countdowns now accurately reflect the quote’s remaining validity.
  * Expired quotes are prevented from proceeding through swap validation.

### CODE REVIEW

File: core/ui/vault/swap/queries/useSwapValidationQuery.test.ts [hunk 1]
```diff
@@ -98,3 +99,26 @@ describe('getImmediateSwapValidationError', () => {
     ).toBe(null)
   })
 })
+
+describe('getSwapQuoteExpiryValidationError', () => {
+  const quoteExpiringAt = (expiresAt: number) =>
+    ({ expiresAt }) as Parameters<typeof getSwapQuoteExpiryValidationError>[0]
```
[the remark this answers is not in the document]
Comment (Ehsan-saradar) [reply core/ui/vault/swap/queries/useSwapValidationQuery.test.ts]: Fixed. The helper now takes `quote: Pick<BoundSwapQuote, 'expiresAt'>` — it reads nothing else, so the narrower type is honest about what it needs and the fixture is a plain `{ expiresAt }` object with no assertion.

The remaining `as` casts in this file (`walletCore: {} as never`, the coin-key fixtures) predate the PR and are untouched.

File: core/ui/vault/swap/queries/useSwapValidationQuery.ts [hunk 2]
```diff
@@ -40,6 +41,18 @@ export const getSwapBalanceValidationError = ({
...
+ * Rejects a quote whose absolute expiry has already passed. React Query keeps a
+ * quote for a day and retains it when a refetch fails, so a cached one can be
+ * displayed long after it was fetched — returning to an unchanged pair and
+ * amount while offline is enough. Nothing else in the swap flow compares a
+ * quote against the clock, so without this the user can submit a stale price.
+ */
+export const getSwapQuoteExpiryValidationError = (
+  quote: BoundSwapQuote,
+  now: number = Date.now()
+): string | null => (quote.expiresAt <= now ? t('swap_quote_expired') : null)
```
[the remark this answers is not in the document]
Comment (Ehsan-saradar) [reply core/ui/vault/swap/queries/useSwapValidationQuery.ts]: Done, with one change from the suggestion: `now` is required rather than defaulted.

The default `Date.now()` was hiding a real bug that your other comment caught — the helper reading the clock itself made it look time-aware when nothing ever re-invoked it. The caller now passes `now` from `useSwapQuoteExpiryClock`, which makes the dependency explicit and is what lets the memo recompute at the expiry instant.

File: core/ui/vault/swap/queries/useSwapValidationQuery.ts [hunk 3]
```diff
@@ -89,8 +102,11 @@ export const useSwapValidationQuery = () => {
       balance: balanceQuery,
       swapQuote: firmSwapQuoteQuery,
     },
-    joinData: ({ balance }) => {
-      return getSwapBalanceValidationError({ amount, balance })
+    joinData: ({ balance, swapQuote }) => {
+      return (
+        getSwapBalanceValidationError({ amount, balance }) ??
+        getSwapQuoteExpiryValidationError(swapQuote)
```
[the remark this answers is not in the document]
Comment (Ehsan-saradar) [reply core/ui/vault/swap/queries/useSwapValidationQuery.ts]: Confirmed and fixed — thanks, this was the real bug in the PR.

Verified the mechanism: `useCombineQueries` wraps its result in `useMemo` keyed on `[eager, joinData, queries]` (`lib/ui/query/hooks/useCombineQueries.ts:84-105`). A quote sitting past its expiry produces no new data, so the memo held and the last `null` verdict stood indefinitely.

Fixed with `useSwapQuoteExpiryClock`, which supplies a `now` that advances at exactly the expiry instant. That changes `joinData`'s identity, so the memo recomputes and Continue blocks.

One timer per quote rather than a poll or a reactive clock — the only interesting moment is known in advance from `expiresAt`. `getSwapQuoteExpiryDelay` is split out so the scheduling decision is unit-tested without fake timers (live / expired / exact-instant).

I did not add a submit-time recheck: the SDK already rejects stale quotes during preparation (`SwapQuoteExpiredError`), so that path has a backstop, and the gap this PR is closing was the UI presenting an expired quote as usable.

**Review by johnnyluo — APPROVED** [review]
LGTM
